### ISSUE TITLE
[feature request] loading font family support

### ISSUE THREAD

**Issue by andykais** [DjDeveloperr/deno-canvas#17]
deno-canvas supports writing text to a canvas:
```ts
import { createCanvas } from 'https://deno.land/x/canvas/mod.ts'

const canvas = createCanvas(500,600)
const ctx = canvas.getContext('2d')
ctx.fillStyle='red'
ctx.fillText(50,50,"Hello World")
await Deno.writeFile("image.png", canvas.toBuffer());
```
this is currently only limited to fonts your system knows about (it might even be more limited than that). Canvas kit has an api for loading custom fonts https://skia.org/docs/user/modules/quickstart/#text-shaping. Itd be great if deno-canvas supported loading custom fonts. I think pulling in the whole paragraph builder api might be substantial, but all I would personally be interested in is mirroring the browser's canvas apis with the addition of being able to load custom fonts. E.g.

```ts
import { createCanvas, loadFont } from 'https://deno.land/x/canvas/mod.ts'

const canvas = createCanvas(500,600)
const ctx = canvas.getContext('2d')
const localFont = await loadFont('./my-fonts/comic-sans.ttf')
ctx.font = localFont
ctx.fillStyle='red'
ctx.fillText(50,50,"Hello World")
await Deno.writeFile("image.png", canvas.toBuffer());
```

**Comment by andykais** [comment]
[edit]
it seems like `ctx.font = ` is not a supported api at all in the current version. Setting font to anything, including fonts available to the system looks like:
```
> ctx.font = 'Roboto'
Uncaught TypeError: Cannot read property 'family' of null
    at q.set [as font] (https://deno.land/x/canvas@v1.3.0/src/lib.js:2264:37)
    at <anonymous>:2:10
```

**Comment by DjDeveloperr** [comment]
For loading custom fonts, there is `canvas.loadFont`.

For setting font you'll need to specify size too.

![image](https://user-images.githubusercontent.com/43033058/134315589-08102ea5-275a-41b0-9e05-6c4e8d0b97ae.png)

**Comment by andykais** [comment]
oh if this is already supported thats fantastic. I am looking at the code now https://github.com/DjDeveloperr/deno-canvas/blob/master/src/types.ts#L1065, what should the descriptors field be?

**Comment by andykais** [comment]
ah, I figured it out:
```ts
canvas.loadFont(fontBuffer, {
  family: 'Comic Sans',
  style: 'normal',
  weight: 'normal',
  variant: 'normal'
})
```
it probably wouldnt hurt to make the type signatures more accurate than `Record<string, string>`. Perhaps:
```ts
type FontDescriptors = {
  /* identifying name of font */
  family: string
  style: 'normal' | 'italic'
  variant: 'normal' | ...
  weight: 'normal' | 'bold' | ...
}
  loadFont(
    bytes: ArrayBuffer | Uint8Array,
    descriptors: FontDescriptors,
  ): void;
```

**Comment by andykais** [comment]
seems like the `context.measureText` results are very inaccurate. 

```ts
const text = "Hello There"
// load a font
    const font = await Deno.readFile(family)
    const font_identifier = new Date().toString()
    canvas.loadFont(font, {
      family: font_identifier
    })
    context.font = `${size}px ${font_identifier}`
// get the font measurements
  const metrics = context.measureText(text)
// draw a rect around it
    context.fillStyle = 'white'
    context.fillRect(0, 0, metrics.width, metrics.fontBoundingBoxAscent + metrics.actualBoundingBoxDescent)
// draw the text
    context.fillStyle =  "black"
    context.fillText(text_chunk, 0, 0)
```

**Comment by andykais** [comment]
if this looks like a real issue I can create a separate issue or change the title on this one

**Comment by andykais** [comment]
Here are some repros. First a jsfiddle which correctly measures the width of text:
https://jsfiddle.net/8dk71toq/2/

<img width="516" alt="Screen Shot 2021-09-24 at 1 08 07 PM" src="https://user-images.githubusercontent.com/6627147/134714188-36a7dadd-de2c-4849-a6d3-5dddde6ec517.png">


and second, deno-canvas incorrectly measuring the same text:
```ts
import { createCanvas } from 'https://deno.land/x/canvas@v1.3.0/mod.ts'

const canvas = createCanvas(500, 200)
const context = canvas.getContext('2d')

function draw_text(text: string, x: number, y: number) {
  const metrics = context.measureText(text)
  console.log(metrics)
  context.fillStyle = 'red'
  context.fillRect(
    x,
    y,
    metrics.actualBoundingBoxLeft + metrics.actualBoundingBoxRight,
    metrics.fontBoundingBoxAscent,
  )

  context.fillStyle = 'black'
  context.fillText(
    text,
    x + metrics.actualBoundingBoxLeft,
    y + metrics.fontBoundingBoxAscent,
  )
}

context.fillStyle = 'white'
context.fillRect(0, 0, canvas.width, canvas.height)
const font_buffer = await Deno.readFile('./fonts/tangerine/Tangerine-Regular.ttf')
canvas.loadFont(font_buffer, {family: 'Tangerine'})
context.font = '50px Tangerine'
draw_text("Hello World", 50, 50)
await Deno.writeFile('canvas.png', canvas.toBuffer())
```

![canvas](https://user-images.githubusercontent.com/6627147/134714332-45758c2d-4408-4249-9f28-08aeae5c3a24.png)


both examples use this font https://fonts.google.com/specimen/Tangerine?query=tangerine

**Comment by andykais** [comment]
some further debugging shows that the builtin font (`monospace` in the case of my mac) correctly measures the text. `50px monospace` is shown below.
![canvas](https://user-images.githubusercontent.com/6627147/134714587-749f34c8-0890-4445-b914-a56f71b9ec22.png)

loading a different font and specifying it via `context.font` _does_ produce different measurements, so its clear that canvaskit is reading the new font. It just doesnt quite interpret the sizes correctly. Is this an issue I should move upstream to canvaskit?

**Comment by DjDeveloperr** [comment]
All values in `TextMetrics` except `width` were added with a hacky workaround, so they might not be right. And no, actual canvaskit does not even implement `measureText` as they just say to use Paragraph API instead. So this is a issue in deno-canvas only.

**Comment by andykais** [comment]
Got it. Well it is unfortunate that measuring text can't be done with deno canvas. I wanted to build out text captions with borders. I suppose there's an extremely hacky workaround for me where I dump the image data and find the max & min x & y

**Comment by DjDeveloperr** [comment]
Right.. text rendering as a whole is not-so-good with `canvaskit-wasm`. Btw, deno-canvas does expose Skia related APIs, such as Paragraph API if that's any helpful for you. I plan on porting Node.js `skia-canvas` using Deno FFI, if that works out we'd have a more performant and compatible canvas API. Though can't say for sure it'll be a thing, or anytime soon as FFI is limited a lot right now.

**Comment by andykais** [comment]
It would definitely be cool to see an ffi bridge to skia. For what I need though, if the paragraph api is exposed that would probably get me everything I need. Would you mind showing an example of how to use it in deno canvas (I have seen canvaskits docs but I'm not sure it's the same api here)

**Comment by DjDeveloperr** [comment]
Skia related APIs are all exposed in a namespace that is default exported from `mod.ts`.

```ts
import Skia from "https://deno.land/x/canvas@v1.3.0/mod.ts";

// Use Skia.Paragraph, Skia.ParagraphBuilder, etc.
// API should be same as canvaskit.
```

**Comment by andykais** [comment]
@DjDeveloperr could you give me a primer on contributing to this library? Would it be this file https://github.com/DjDeveloperr/deno-canvas/blob/master/src/lib.js#L2750? Its hard to tell if this is a source file because some of this code looks minified. Could you show where another skia specific method is used as an example?

**Comment by DjDeveloperr** [comment]
Yeah, in `src/lib.js`. Right.. it was minified at first. `a` variable contains everything exported as default in `mod.ts`, so all Skia APIs. You can use ctrl+f to find implementation of `measureText`.

**Comment by DjDeveloperr** [comment]
@andykais your first attempt at the implementation looks great. I think it would be good enough for first pass (better than currently inaccurate implementation)

And yes, we do have access to `Font` in context (it's a minified property, which we can access through `context.we`), and even `FontMgr` (in `canvas.Cf`).

So I think you can PR this function (but slightly modified to accept `FontMgr` and `Font` instead of), and I can add it in `lib.js` 🤔

**Comment by andykais** [comment]
whats the current CanvasKit version?

**Comment by DjDeveloperr** [comment]
It is 0.32.0

**Comment by webdev3301** [comment]
I think I've figure out how to measure text correctly.

First, ignore measureText() method. It is not accurate.

```ts

// Make sure to set the font size first!
 context2d.font = `16px sans-serif`;

 // The text to display
 const text: string = `Hello World`;

 // Measure the width of a single character
 const charWidth: number = Math.floor(context2d.measureText("X").width);

 // Get the length of the text string
 const textLength: number = text.length;

 // textWidth at selected fontSize
 const textWidth: number = charWidth * textLength;
```

**Comment by andykais** [comment]
@webdev3301 I believe there is still an issue with measuring custom loaded fonts. See this message above: https://github.com/DjDeveloperr/deno-canvas/issues/17#issuecomment-926166920
